outlines/models/azure.py

```diff
@@ -0,0 +1,119 @@
+"""Integration with Azure OpenAI's API."""
+import functools
+import os
+from dataclasses import replace
+from typing import Optional
+
+from outlines.models.openai import OpenAI, OpenAIConfig
+
+__all__ = ["AzureOpenAI", "azure_openai"]
+
+
+AZURE_API_VERSION = "2023-05-15"
+
+
+class AzureOpenAI(OpenAI):
+    def __init__(
+        self,
+        model_name: str,
+        deployment_name: str,
+        azure_endpoint: Optional[str] = None,
+        api_key: Optional[str] = None,
+        max_retries: int = 6,
+        timeout: Optional[float] = None,
+        system_prompt: Optional[str] = None,
+        config: Optional[OpenAIConfig] = None,
+    ):
+        """Create an `AzureOpenAI` instance.
+
+        Parameters
+        ----------
+        model_name
+            The name of the OpenAI model being used
+        deployment_name
+            The name of your Azure OpenAI deployment
+        api_key
+            Secret key to use with the OpenAI API. One can also set the
+            `OPENAI_API_KEY` environment variable, or the value of
+            `openai.api_key`.
+        max_retries
+            The maximum number of retries when calls to the API fail.
+        timeout
+            Duration after which the request times out.
+        system_prompt
+            The content of the system message that precedes the user's prompt.
+        config
+            An instance of `OpenAIConfig`. Can be useful to specify some
+            parameters that cannot be set by calling this class' methods.
+
+        """
+        try:
+            import openai
+        except ImportError:
+            raise ImportError(
+                "The `openai` library needs to be installed in order to use Outlines' Azure OpenAI integration."
+            )
+        try:
+            client = openai.OpenAI()
```

Comment:
Thanks for adding Azure support! When I try to use azure_openai, it fails at this line of code, since the OpenAI client attempts to authenticate. Is there a different way we can validate the model name, besides instantiating the client?

Reply:
I am now leaning towards "let it crash" i.e. let users enter anything for the model name and then handle a potential error gracefully, letting the user know that the model name is incorrect.

Reply:
I don’t have the time to fix it right now, but happy to review a PR.